airflow/www/security.py

```diff
@@ -578,22 +576,22 @@ def _get_all_non_dag_permissions(self) -> dict[tuple[str, str], Permission]:
         return {
             (action_name, resource_name): viewmodel
             for action_name, resource_name, viewmodel in (
-                self.appbuilder.get_session.query(self.permission_model)
-                .join(self.permission_model.action)
-                .join(self.permission_model.resource)
-                .filter(~self.resource_model.name.like(f"{permissions.RESOURCE_DAG_PREFIX}%"))
-                .with_entities(self.action_model.name, self.resource_model.name, self.permission_model)
-                .all()
+                self.appbuilder.get_session.execute(
+                    select(self.action_model.name, self.resource_model.name, self.permission_model)
+                    .join(self.permission_model.action)
+                    .join(self.permission_model.resource)
+                    .where(~self.resource_model.name.like(f"{permissions.RESOURCE_DAG_PREFIX}%"))
+                ).all()
             )
         }
 
     def _get_all_roles_with_permissions(self) -> dict[str, Role]:
         """Returns a dict with a key of role name and value of role with early loaded permissions."""
         return {
             r.name: r
-            for r in self.appbuilder.get_session.query(self.role_model).options(
-                joinedload(self.role_model.permissions)
-            )
+            for r in self.appbuilder.get_session.scalars(
+                select(self.role_model).options(joinedload(self.role_model.permissions))
+            ).unique()
```

Comment:
Is this correct? There wasn’t a unique call previously (although it probably doesn’t matter since this is also immediately turned into a set so duplicates are eliminated anyway)

Reply:
Why did the previous method not have this issue?

Reply:
one is a ScalarResult and the other is a orm query. May be the ScalarResult is treated as collection and the query is not
```
(Pdb) type(self.appbuilder.get_session.scalars(select(self.role_model).options(joinedload(self.role_model.permissions))))
<class 'sqlalchemy.engine.result.ScalarResult'>
(Pdb) type(self.appbuilder.get_session.query(self.role_model).options(joinedload(self.role_model.permissions)))
<class 'sqlalchemy.orm.query.Query'>
```

Reply:
Looks like the query API automatically deduplicates on primary key:

https://docs.sqlalchemy.org/en/20/faq/sessions.html#my-query-does-not-return-the-same-number-of-objects-as-query-count-tells-me-why

> The Query object, when asked to return a list of ORM-mapped objects, will deduplicate the objects based on primary key.

For this particular purpose `unique()` is fine since duplicated rows would have the same hash, but alternatively `unique(lambda r: r.id)` is probably both more accurate and slightly faster.